### PR TITLE
Fixed rare bulk_update() deadlock in ContentSaver.

### LINKED ISSUE #2430: bulk_update() in content-stages can cause (very rare) deadlock
**Version**
3.14

**Describe the bug**
In high-concurrency environments, with overlapping content, calling bulk_update() can cause a deadlock. Specifically, this call:

https://github.com/pulp/pulpcore/blob/main/pulpcore/plugin/stages/content_stages.py#L158-L164

Ordering the list-to-be-updated does not, alas, protect us - because Postgres doesn't guarantee order when doing an update like this.

**To Reproduce**
We have only seen this "in the wild" once, syncing 8-10 repos with similar content at the same time with 10 workers available.

**Expected behavior**
Don't deadlock.

**Additional context**
This is the traceback from the initial description for

https://bugzilla.redhat.com/show_bug.cgi?id=2062526

We fixed the deadlock noted in https://bugzilla.redhat.com/show_bug.cgi?id=2062526#c2 under #2420

### PR DESCRIPTION
**Pull request by ggainey**
See https://bugzilla.redhat.com/show_bug.cgi?id=2062526 for details on
forcing the deadlock and discussion of the fix.

fixes #2430.
[nocoverage]

### CODE REVIEW

**Comment by dralley** [comment]
Does this need a 3.14 backport?

**Comment by ggainey** [comment]
> Does this need a 3.14 backport? (Presumably a manual one)

Target is set to 6.11 (which is 3.16) - but 6.10.z flag is also set, so we prob should bring it back to 3.14 as well.

**Review by dralley — APPROVED** [review]

File: pulpcore/plugin/stages/content_stages.py [hunk 1]
```diff
@@ -155,13 +155,39 @@ def process_batch():
...
+                            .order_by("pulp_id")
+                            .select_for_update()
+                            .values_list()
+                        )
+                        ContentArtifact.objects.bulk_update(to_update_ca_bulk, ["artifact"])
 
-                    ContentArtifact.objects.bulk_update(to_update_ca_bulk, ["artifact"])
+                    # To avoid a similar deadlock issue when calling get_or_create, we sort the
+                    # "new" CAs to make sure inserts happen in a defined order. Since we can't
+                    # trust the pulp_id (by the time we go to create a CA< it may already exist,
```
Comment (dralley) [inline pulpcore/plugin/stages/content_stages.py:188]: typo: s/</,

Comment (ggainey) [reply pulpcore/plugin/stages/content_stages.py:188]: > typo: s/</,

Ugh, prob my most-common typo :(

**Review by mdellweg — APPROVED** [review]

**Pull request merged**